docs/CppInteroperability/ForwardVision.md

```diff
@@ -0,0 +1,266 @@
+
+## Assumptions
+
+C++ and Swift interoperability assumes that projections only appear in the form of a pointer type being returnred from a method. While this is not the only way to create such a projection, it is by far the most common. In order to maintain C and Objective-C interoperability and allow Swift programmers to use C++ APIs, these edge cases must still be imported even if they could result in a less safe program (API authors already have tools to mark these APIs as unavailable in Swift, which they are encouraged to use). These are two of the most prominent edge cases:
+
+```
+int *begin(std::vector<int> *v) { return v->data(); }
+
+struct VectorLike {
+  // ...
+  void begin(int **out) { *out = data(); }
+};
+```
+
+This assumption is in line with other assumptions the Swift compiler makes when importing C++ APIs (such as methods on trivial types will not return stack pointers and passing to a const reference will not mutate the pointee object).
+
+
+## Lifetime and safety of owned types
+
+Value types that own memory do not natively exist in Swift today. Because Swift was not built around this kind of value type, Swift handles the lifetime of value types in a subtly different way from C++. This subtle difference makes a naive mapping of C++ value types that own memory to native Swift value types extremely dangerous, especially when dealing with projections of owned storage. Let’s look at an example Swift program that naively imports some owned type and returns a projections of it:
+
+```
+var v = vector(1)
+let start = v.begin()
+doSomething(start)
+fixLifetime(v)
+```
+
+To understand the problem with this code, the following snippet highlights where an implicit copy is created and destroyed:
+
+```
+var v = vector(1)
+let copy = copy(v)
+let start = copy.begin()
+destroy(copy)
+doSomething(start)
+fixLifetime(v)
+```
+
+Here, because Swift copies `v` into a temporary with a tight lifetime before the call to `begin`, `v` projects a dangling reference. This is an example of how subtly different lifetime models make using C++ types from Swift hard, if their semantics aren’t understood by the compiler.
+
+To make these APIs safe and usable, Swift cannot import unsafe projections of types that own memory, because they don’t fit the Swift model. Instead, the Swift compiler can try to infer what, semantically, the API is trying to do, or the library author can provide this information via annotations. In this case, the Swift compiler can infer that begin returns an iterator, which Swift can represent through the existing, safe Swift iterator interface. In the example above, “start” is a pointer type. Using this pointer returned by the “begin” method is unsafe, but the type of start itself is not unsafe. In other words, safety restrictions need not be applied to pointer types themselves but rather their unsafe uses.
```

Comment:
Is it counter to our goals to do things like _change_ lifetime behaviors for imported C++ types to make them behave a little more C++ like? I.e., use lexical scoping rules for imported C++ types here.

Reply:
Let me start off by saying I think that's a good idea (in this specific case). However, more generally, I think that's a slippery slope. There's always a tradeoff when making imported APIs behave less like "Swift". It may make it easier to use the APIs, but as we lose Swift semantics, which creates cognitive burden for the programmer. Drifting away from Swift semantics also means programers might not get a lot of the benefit of using Swift. We don't want this to feel like C++ with different syntax, we want it to feel like Swift. 

That is all to say, I think we need to be very selective in where we do this.

Reply:
Yeah, I see the difficulty. On the one hand, working with C++ APIs is going to require _some_ knowledge of C++ semantics. Imported Objective-C classes today exhibit behavior that cannot really be explained in terms of Swift semantics and I expect it will be the same for imported C++ APIs. On the other hand there are also invariants of Swift that we will likely want to make sure get preserved even for imported C++ APIs.